File: backend/app/db/models/receiving.py
from datetime import date
from typing import TYPE_CHECKING

from sqlalchemy import BigInteger, Date, ForeignKey, Integer, String, text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.db.models.logistics import Shipment
    from app.db.models.product import ProductVariant
    from app.db.models.warehouse import Warehouse


class Receiving(Base):
    __tablename__ = "receiving"

    receiving_id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    # Unique: the diagram's `SHIPMENT ||--o| RECEIVING` cardinality means a
    # shipment has at most one receiving record.
    shipment_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("shipments.shipment_id"), unique=True, nullable=False
    )
    warehouse_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("warehouses.warehouse_id"), nullable=False
    )
    received_date: Mapped[date] = mapped_column(Date, nullable=False)
    status: Mapped[str] = mapped_column(String, nullable=False)

    shipment: Mapped["Shipment"] = relationship(back_populates="receiving")
    warehouse: Mapped["Warehouse"] = relationship(back_populates="receivings")
    items: Mapped[list["ReceivingItem"]] = relationship(back_populates="receiving")


class ReceivingItem(Base):
    __tablename__ = "receiving_items"

    receiving_item_id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    receiving_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("receiving.receiving_id"), nullable=False
    )
    variant_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("product_variants.variant_id"), nullable=False
    )
    quantity_expected: Mapped[int] = mapped_column(Integer, nullable=False)
    quantity_received: Mapped[int] = mapped_column(
        Integer, nullable=False, server_default=text("0")
    )

    receiving: Mapped["Receiving"] = relationship(back_populates="items")
    variant: Mapped["ProductVariant"] = relationship(back_populates="receiving_items")

File: frontend/src/data/fixtures/receiving.ts
import type { Receiving, ReceivingItem } from "@/data/types";

export const receivings: Receiving[] = [
  { receivingId: "RC-000", shipmentId: "SH-000", warehouseId: "WH-2", receivedDate: "2026-08-06", status: "completed" },
  { receivingId: "RC-001", shipmentId: "SH-003", warehouseId: "WH-2", receivedDate: "2026-08-09", status: "completed" },
  { receivingId: "RC-002", shipmentId: "SH-004", warehouseId: "WH-1", receivedDate: "2026-08-11", status: "completed" },
];

export const receivingItems: ReceivingItem[] = [
  { receivingItemId: "RI-000-1", receivingId: "RC-000", productId: "PROD-1", quantityExpected: 60, quantityReceived: 60 },
  { receivingItemId: "RI-000-2", receivingId: "RC-000", productId: "PROD-3", quantityExpected: 20, quantityReceived: 20 },

  { receivingItemId: "RI-001-1", receivingId: "RC-001", productId: "PROD-4", quantityExpected: 30, quantityReceived: 30 },
  { receivingItemId: "RI-001-2", receivingId: "RC-001", productId: "PROD-8", quantityExpected: 15, quantityReceived: 15 },
  { receivingItemId: "RI-001-3", receivingId: "RC-001", productId: "PROD-5", quantityExpected: 10, quantityReceived: 10 },

  { receivingItemId: "RI-002-1", receivingId: "RC-002", productId: "PROD-2", quantityExpected: 60, quantityReceived: 60 },
  { receivingItemId: "RI-002-2", receivingId: "RC-002", productId: "PROD-9", quantityExpected: 45, quantityReceived: 45 },
];
